docs/proposals/20210614-podunavailablebudget.md

```diff
@@ -0,0 +1,148 @@
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [API Definition](#api-definition)
+    - [Implement](#implement)
+    - [Program Comparison](#program-comparison)
+  - [User Cases](#user-cases)
+    - [A Sample Case](#a-sample-case)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+In kubernetes, powerful workloads (e.g., deployment, statefulset, etc.) are often used to define and manage applications. It not only manages the number of pods, but also provides powerful rolling capabilities such as MaxUnavailable, Partition, etc. However, workload can only help the application reach the desired number of pods as fast as possible, but does not guarantee that the number of available pods will remain at the desired value at any given moment. In the following voluntary disruption scenarios, there are still business disruption and SLA degradation situations:
+1. When the owner deletes the workload by mistake, all the associated pods will be deleted
+2. When upgrading the application version, all pods are restarted at the same time due to the wrong configuration of the rolling upgrade policy, for example MaxUnavailable=0
+3. Many pods are evicted when the cluster administrator upgrades the linux kernel
+
+[Pod Disruption Budget](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/) is an attempt to achieve the application protection capability through EvictionREST interface for voluntary disruption scenario. This solution can achieve a good effect of security protection in Pod Eviction scenario, but it does not achieve the application protection capability for scenarios such as rolling upgrade, mistaken workload deletion.
+
+## Proposal
+PodUnavailableBudget is an attempt for voluntary disruption scenario, which is not only compatible with Kubernetes PDB protection for Pod Eviction scenario, but also able to support the protection ability of various scenarios such as k8s workload native rolling upgrade, kruise workload in-place upgrade and mistaken workload deletion, etc.
+### API Definition
+```
+// PodUnavailableBudgetSpec defines the desired state of PodUnavailableBudget
+type PodUnavailableBudgetSpec struct {
+	// Selector label query over pods managed by the budget
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// TargetReference contains enough information to let you identify an workload for PodUnavailableBudget
+	// Selector and TargetReference are mutually exclusive, TargetReference is priority to take effect
+	TargetReference *TargetReference `json:"targetRef,omitempty"`
```

Comment:
please comments about when target reference is required when we already have selector fields